README.md
# VisualStructsPy

Graphically displays data structures such as arrays, lists, and trees in Python 3 using PyQt5 wrapper for Qt5.                                                    

#### stable release

This branch (master) contains code which *should* interpret just fine, and does not contain any incomplete features 
(currently empty until first release from 'devel').  

#### development

To view the current state of development, visit the '[devel](https://github.com/chrismabon/VisualStructsPy/tree/devel)' branch.
Be warned, this branch may contain incomplete features, bugs, and/or may not interpret at all. 

#### public domain code

VisualStructsPy is being developed without any licensing restrictions or copyright: it is public domain code and 
completely free for use by anyone for any purpose.

```
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```  

CHANGELOG.md
# Change Log

## 2019-10-23

##### Changes

- Removed license and released project into public domain
- Created branch 'devel' from 'master'
- Cleaned 'master' branch of obsolete code

## 2019-08-13

##### Changes

- Re-licensed to Duo Extrema software

## 2019-08-10

##### Additions

- Add rudimentary array-graphing

## 2019-08-01

#### initial commit, v0.1.0

src/canvas.py
import sys
import src.vsArray as Array

from PyQt5.QtWidgets import (QWidget, QLabel, QComboBox, QGridLayout, QFrame, QSplitter, QMainWindow,
                             QApplication, QVBoxLayout, QHBoxLayout, QGraphicsScene, QGraphicsView, QAction,
                             qApp)
from PyQt5.QtGui import (QPainter, QColor, QFont, QIcon)
from PyQt5.QtCore import (Qt, QPoint, QRect)


class Gui(QMainWindow):
    def __init__(self):
        super().__init__()

        self.guiX = 100
        self.guiY = 100
        self.guiWidth = 1000
        self.guiHeight = 600
        self.guiMargins = 50

        self._initUI()

    def _initUI(self):
        exitAct = QAction(QIcon('exit.png'), '&Exit', self)
        exitAct.setShortcut('Ctrl+Q')
        exitAct.setStatusTip('Exit application')
        exitAct.triggered.connect(qApp.quit)

        self.statusBar()

        menubar = self.menuBar()
        fileMenu = menubar.addMenu('&File')
        fileMenu.addAction(exitAct)

        self._setArrProps()

        self.setGeometry(self.guiX, self.guiY, Array.VSArray.props['totalWidth'], Array.VSArray.props['totalHeight'])
        self.setWindowTitle('Visual Structs')
        self.show()

    def _longestStr(self, data):
        i = 0
        longest = 0
        while i < data.__len__():
            if longest < data[i].__len__():
                longest = data[i].__len__()
            i += 1

        return longest

    def _setArrProps(self):
        Array.VSArray.props['numElems'] = len(Array.VSArray.data)
        Array.VSArray.props['textMargins'] = int(Array.VSArray.props['fontSize'] * 0.5)
        Array.VSArray.props['elemWidth'] = (((Array.VSArray.props['fontSize'] * 0.85)
                                             * self._longestStr(Array.VSArray.data))
                                            + (Array.VSArray.props['textMargins'] * 10))
        Array.VSArray.props['elemHeight'] = Array.VSArray.props['fontSize'] * 2
        Array.VSArray.props['totalHeight'] = (self.guiMargins * 2) + (Array.VSArray.props['elemHeight']
                                                                      * Array.VSArray.props['numElems'])
        Array.VSArray.props['totalWidth'] = (self.guiMargins * 2) + Array.VSArray.props['elemWidth']

    def paintEvent(self, event):
        qp = QPainter()
        qp.begin(self)
        self._drawArray(qp)
        qp.end()

    def _drawArray(self, qp):
        qp.setFont(QFont(Array.VSArray.props['fontFamily'], Array.VSArray.props['fontSize']))
        elemWidth = Array.VSArray.props['elemWidth']
        elemHeight = Array.VSArray.props['elemHeight']
        margins = self.guiMargins

        def _drawArrElem(x, y, w, h, index, data):
            qp.setPen(QColor(0, 0, 0))
            qp.setBrush(QColor(150, 250, 250))
            qp.drawRect(x, y, w, h)
            margins = Array.VSArray.props['textMargins']
            padding = Array.VSArray.props['fontSize']
            qp.drawText(x + margins, y + margins + (elemHeight / 2), str(index))
            qp.drawLine(x + margins + (padding * 2), y, x + margins + (padding * 2),
                        y + margins + elemHeight)
            qp.drawText(x + margins + padding + (Array.VSArray.props['fontSize'] * 2),
                        y + margins + (elemHeight / 2), str(data))

        i = 0
        while i < Array.VSArray.props['numElems']:
            offset = i * elemHeight
            _drawArrElem(margins, margins + offset, elemWidth, elemHeight, i, Array.VSArray.data[i])
            i += 1


if __name__ == '__main__':
    app = QApplication(sys.argv)
    ex = Gui()
    sys.exit(app.exec_())


src/vsArray.py
class VSArray:
    data = [
        'Donec non justo auctor, efficitur augue sit amet, semper dolor.',
        'Maecenas in fermentum dolor. Vestibulum porttitor volutpat odio, aliquam suscipit erat volutpat vel. ',
        'Nulla ac tellus lacus.',
        'Morbi euismod mi lorem, non placerat tellus iaculis et.',
        'Ut in feugiat nisl, in fringilla leo.',
        'In hendrerit blandit velit vitae ullamcorper.',
        'Mauris varius quam vitae odio sodales, non faucibus arcu mattis.',
        'Nullam a pretium ex, in hendrerit nulla.',
        'Donec fermentum purus eget magna fermentum blandit.',
        'Vestibulum mollis tortor consectetur diam commodo, sagittis gravida diam laoreet.',
        'Nulla id neque vel diam vestibulum convallis id a felis.',
        'Quisque ac eros at magna mollis suscipit in vitae turpis.',
        'Duis sodales mattis scelerisque.',
        'Phasellus lectus mi, blandit eu feugiat vitae, viverra vel orci.'
    ]
    props = {
        'numElems': 0,
        'elemWidth': 0,
        'elemHeight': 0,
        'totalHeight': 0,
        'totalWidth': 0,
        'textMargins': 0,
        'fontSize': 14,
        'fontFamily': 'Monospace'
    }

    def __init__(self):
        pass


VisualStructsPy/VSList.py
from VisualStructsPy.VSNode import VSNode

from PyQt5.QtGui import (QColor, QFont, QTransform)


class VSList:
    def __init__(self):
        self.font = {
            'fontSize': 20,
            'fontFamily': 'Monospace'
        }
        self.data = [
            'Nulla ac tellus lacus.',
            'Morbi euismod mi lorem, non placerat tellus iaculis et.',
            'Ut in feugiat nisl, in fringilla leo.',
            'In hendrerit blandit velit vitae ullamcorper.',
            'Nullam a pretium ex, in hendrerit nulla.',
            'Donec fermentum purus eget magna fermentum blandit.',
            'Nulla id neque vel diam vestibulum convallis id a felis.',
            'Quisque ac eros at magna mollis suscipit in vitae turpis.',
            'Duis sodales mattis scelerisque.'
        ]
        self.props = {
            'numElems': 0,
            'elemWidth': 0,
            'elemHeight': 0,
            'elemMargin': 0,
            'totalHeight': 0,
            'totalWidth': 0,
            'textMargins': 0,
            'padding': self.font['fontSize'] / 2
        }

    def drawList(self, qp):
        elemWidth = self.props['elemWidth']
        elemHeight = self.props['elemHeight']
        padding = self.props['padding']
        txtMargins = self.props['textMargins']
        qp.setFont(QFont(self.font['fontFamily'], self.font['fontSize']))
        qp.setPen(QColor(0, 0, 0))

        def _drawListElem(x, y, w, h, data):
            qp.setBrush(QColor(150, 250, 250))
            qp.setFont(QFont(self.font['fontFamily'], self.font['fontSize'] - 6))
            qp.drawEllipse(x + txtMargins + (elemWidth / 2) + padding, y + padding, w, h)
            qp.drawText(x + (txtMargins * 2) + padding + (elemWidth / 2), y + txtMargins + padding
                        + (elemHeight / 2.5), str(data))

        def _drawListHead(count):
            qp.setBrush(QColor(250, 0, 0))
            x = self.props['elemMargin']
            y = self.props['elemMargin']
            w = (self.props['totalWidth'] / 2.5) - (self.props['elemMargin'])
            h = (self.props['totalHeight'] / 2) - (self.props['elemMargin'] * 2)
            qp.drawRect(x, y, w, h)
            qp.drawText(x + txtMargins, y - txtMargins, 'Count:' + str(count))
        _drawListHead(self.data.__len__())

        i = 0

        while i < self.props['numElems']:
            transform = QTransform.translate(QTransform(1, 0, 0, 1, i * 5.0, 20.0), 1.0, 1.0)
            qp.setTransform(transform)
            offset = int(i * elemWidth + (elemWidth / 2))
            _drawListElem(txtMargins + offset + (padding * i * 2), txtMargins + elemHeight, elemWidth, elemHeight,
                          self.data[i][0:4])
            i += 1


VisualStructsPy/VSArray.py
from PyQt5.QtGui import (QColor, QFont)


class VSArray:
    font = {
        'fontSize': 14,
        'fontFamily': 'Monospace'
    }
    data = [
        'Donec non justo auctor, efficitur augue sit amet, semper dolor.',
        'Maecenas in fermentum dolor. Vestibulum porttitor volutpat odio, aliquam suscipit erat volutpat vel. ',
        'Nulla ac tellus lacus.',
        'Morbi euismod mi lorem, non placerat tellus iaculis et.',
        'Ut in feugiat nisl, in fringilla leo.',
        'In hendrerit blandit velit vitae ullamcorper.',
        'Mauris varius quam vitae odio sodales, non faucibus arcu mattis.',
        'Nullam a pretium ex, in hendrerit nulla.',
        'Donec fermentum purus eget magna fermentum blandit.',
        'Vestibulum mollis tortor consectetur diam commodo, sagittis gravida diam laoreet.',
        'Nulla id neque vel diam vestibulum convallis id a felis.',
        'Quisque ac eros at magna mollis suscipit in vitae turpis.',
        'Duis sodales mattis scelerisque.',
        'Phasellus lectus mi, blandit eu feugiat vitae, viverra vel orci.'
    ]
    props = {
        'numElems': 0,
        'elemWidth': 0,
        'elemHeight': 0,
        'totalHeight': 0,
        'totalWidth': 0,
        'textMargins': 0,
        'padding': font['fontSize'] / 2
    }

    def __init__(self):
        pass

    def _drawArray(self, qp):
        qp.setFont(QFont(self.font['fontFamily'], self.font['fontSize']))
        elemWidth = self.props['elemWidth']
        elemHeight = self.props['elemHeight']
        txtMargins = self.props['textMargins']

        def _drawArrElem(x, y, w, h, index, data):
            qp.setPen(QColor(0, 0, 0))
            qp.setBrush(QColor(150, 250, 250))
            qp.drawRect(x, y, w, h)
            padding = self.font['fontSize']
            qp.drawText(x + txtMargins, y + txtMargins + (elemHeight / 2), str(index))
            qp.drawLine(x + txtMargins + (padding * 2), y, x + txtMargins + (padding * 2),
                        y + txtMargins + elemHeight)
            qp.drawText(x + txtMargins + padding + (self.font['fontSize'] * 2),
                        y + txtMargins + (elemHeight / 2), str(data))

        i = 0
        while i < self.props['numElems']:
            offset = i * elemHeight
            _drawArrElem(txtMargins, txtMargins + offset, elemWidth, elemHeight, i, self.data[i])
            i += 1



VisualStructsPy/VSCanvas.py
"""@package VisualStructsPy
Graphically displays data structures such as arrays, lists, and trees.

Users simply pass a list of data elements to the appropriate constructor function to instantiate and
populate a structure with that data. Note that trees have additional arguments to determine the manner
in which nodes are added.
"""

import sys
import VisualStructsPy.VSArray as VSArray
import VisualStructsPy.VSList as VSList
import VisualStructsPy.VSTree as VSTree

from PyQt5.QtWidgets import (QMainWindow, QApplication, QAction, qApp)
from PyQt5.QtGui import (QPainter, QColor, QFont, QIcon)


class Canvas(QMainWindow):
    def __init__(self):
        super().__init__()

        self.vsArray = VSArray.VSArray()
        self.vsList = VSList.VSList()
        self.vsTree = VSTree.VSTree()

        self.guiX = 100
        self.guiY = 100
        self.guiWidth = 1000
        self.guiHeight = 600
        self.guiMargins = 50

        self._initUI()

    def _initUI(self):
        exitAct = QAction(QIcon('exit.png'), '&Exit', self)
        exitAct.setShortcut('Ctrl+Q')
        exitAct.setStatusTip('Exit application')
        exitAct.triggered.connect(qApp.quit)

        self.statusBar()

        menubar = self.menuBar()
        fileMenu = menubar.addMenu('&File')
        fileMenu.addAction(exitAct)

        # self._setArrProps()
        self._setListProps()

        self.setGeometry(self.guiX, self.guiY,self.guiWidth, self.guiHeight)
        self.setWindowTitle('Visual Structs')
        self.show()

    def setGuiWidth(self, width):
        self.guiWidth = width

    def setGuiHeight(self, height):
        self.guiHeight = height

    def _longestStr(self, data):
        i = 0
        longest = 0
        while i < data.__len__():
            if longest < data[i].__len__():
                longest = data[i].__len__()
            i += 1

        return longest

    def _setArrProps(self):
        self.vsArray.props['numElems'] = len(self.vsArray.data)
        self.vsArray.props['textMargins'] = int(self.vsArray.font['fontSize'] * 0.5)
        self.vsArray.props['elemWidth'] = (((self.vsArray.font['fontSize'] * 0.85)
                                            * self._longestStr(self.vsArray.data))
                                           + (self.vsArray.props['textMargins'] * 10))
        self.vsArray.props['elemHeight'] = self.vsArray.font['fontSize'] * 2
        self.vsArray.props['totalHeight'] = (self.guiMargins * 4) + (self.vsArray.props['elemHeight']
                                                                        * self.vsArray.props['numElems'])
        self.vsArray.props['totalWidth'] = (self.guiMargins * 4) + self.vsArray.props['elemWidth']
        self.guiWidth = self.vsArray.props['totalWidth']
        self.guiHeight = self.vsArray.props['totalHeight']

    def _setListProps(self):
        self.vsList.props['numElems'] = len(self.vsList.data)
        self.vsList.props['textMargins'] = int(self.vsList.font['fontSize'] * 0.5)
        self.vsList.props['elemWidth'] = (self.vsList.font['fontSize'] * 4) + (self.vsList.props['textMargins'] * 0.5)
        self.vsList.props['elemHeight'] = self.vsList.font['fontSize'] * 4
        self.vsList.props['totalHeight'] = (self.guiMargins * 2) + ((self.vsList.props['elemHeight'] * 2)
                                                                      + self.vsList.props['elemMargin'])
        self.vsList.props['totalWidth'] = (self.guiMargins * 2) + ((self.vsList.props['elemWidth']
                                                                      + (self.vsList.props['elemMargin'] * 2)
                                                                    + (self.vsList.props['padding'] * 2))
                                                                     * self.vsList.props['numElems'])
        self.guiWidth = self.vsList.props['totalWidth']
        self.guiHeight = self.vsList.props['totalHeight']
        self.vsList.props['elemMargin'] = self.guiWidth / self.vsList.props['numElems']

    def paintEvent(self, event):
        qp = QPainter()
        qp.begin(self)
        # self._drawArray(qp)
        self.vsList.drawList(qp)
        qp.end()


if __name__ == '__main__':
    app = QApplication(sys.argv)
    ex = Canvas()
    sys.exit(app.exec_())
